### ISSUE TITLE
test relative rather than absolute error in SYCL nstream

### ISSUE THREAD

**Issue by jeffhammond** [ParRes/Kernels#407]
From @illuhad:
```
./nstream-sycl 100 50000000
Parallel Research Kernels version 2.16
C++11/SYCL STREAM triad: A = B + scalar * C
Number of iterations = 100
Vector length        = 50000000
Offset               = 0
Skipping host device since it is too slow for large problems
Failed Validation on output array
       Expected checksum: 4.04e+10
       Observed checksum: 4.04e+10
ERROR: solution did not validate
Solution validates
64B Rate (MB/s): 765353 Avg time (s): 0.00209054
```
> When the vector length is increased, the single precision result fails to
> validate, which I suppose is due to single precision rounding errors becoming
> relevant. Maybe we should increase the tolerance for SP, at least for very
> large problems?

**Comment by rfvander** [comment]
All floating point operands and results can be represented exactly, since they only assume integral values. The failure to validate must come from somewhere else, not rounding. I would resist the temptation to change the error threshold and try an identify the actual source of the error. Maybe run the traditional SERIAL C version with the same parameters.

Rob

**Comment by jeffhammond** [comment]
@rfvander I need to debug the expression `(std::fabs(ar-asum)/asum > epsilon)`.  I fixed it already, but there was a time when the error check in p2p would pass when the result was NAN, because of the rules of FPE.

**Comment by rfvander** [comment]
@jeffhammond Yes, I think I committed that NaN error a long time ago, but I don't remember the specifics. At any rate, the error reported above is not due to NaNs, since both the expected and observed checksums are evidently regular numbers.

**Comment by illuhad** [comment]
No, it's not due to NaNs - printing the numbers with more accuracy reveals that the *expected* value is in fact wrong:
```
./nstream-sycl 100 100000000
Parallel Research Kernels version 2.6
C++11/SYCL STREAM triad: A = B + scalar * C
Number of iterations = 100
Vector length        = 100000000
Offset               = 0
Skipping host device since it is too slow for large problems
Failed Validation on output array
       Expected checksum: 80799997952
       Observed checksum: 80800000000
ERROR: solution did not validate
```

**Comment by illuhad** [comment]
Problem is fixed by using `double` instead of `T` when calculating the expected value.

**Comment by rfvander** [comment]
Excellent, thanks. It will be good to make a sweep over all SYCL implementations to see if similar errors may occur for other kernels.

**Comment by jeffhammond** [comment]
Not just SYCL, anywhere we use float32.

I suspect the accumulation is wrong but the actual results aren’t. I’ve been ignoring these issues already but it’s clearly time to fix them.
